File: src/com/hoshi/rubotoprocessing2/ProcessingSketchActivity.java
package com.hoshi.rubotoprocessing2;

import processing.core.*;
import processing.data.*;
import processing.event.*;
import processing.opengl.*;

import java.io.IOException;
import org.ruboto.Script;
import android.app.ProgressDialog;
import android.os.Bundle;
import android.util.Log;

public class ProcessingSketchActivity extends PApplet implements org.ruboto.RubotoComponent {

  static final public boolean DEBUG = true;
  protected boolean availableSetup = false;
  protected boolean availableDraw = false;
  protected boolean availableKeyPressed = false;
  protected boolean availableKeyReleased = false;
  protected boolean availableMousePressed = false;
  protected boolean availableMouseReleased = false;
  protected boolean availableMouseClicked = false;
  protected boolean availableMouseDragged = false;
  protected boolean availableMouseMoved = false;
  protected boolean availableMouseEntered = false;
  protected boolean availableMouseExited = false;

  private final org.ruboto.ScriptInfo scriptInfo = new org.ruboto.ScriptInfo();
  public org.ruboto.ScriptInfo getScriptInfo() {
      return scriptInfo;
  }

  public void onCreate(android.os.Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    if (org.ruboto.JRubyAdapter.isInitialized()); {
      getScriptInfo().setFromIntent(getIntent());
      org.ruboto.ScriptLoader.loadScript(this);

      String rubyClassName = scriptInfo.getRubyClassName();
      String scriptBase = ".private_instance_methods(false).any?{|m| m.to_sym == ";

      availableSetup = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_setup}");
      availableDraw = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_draw}");
      availableKeyPressed = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_key_pressed}");
      availableKeyReleased = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_key_released}");
      availableMousePressed = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_mouse_pressed}");
      availableMouseReleased = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_mouse_released}");
      availableMouseClicked = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_mouse_clicked}");
      availableMouseDragged = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_mouse_dragged}");
      availableMouseMoved = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_mouse_moved}");
      availableMouseEntered = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_mouse_entered}");
      availableMouseExited = (Boolean)org.ruboto.JRubyAdapter.runScriptlet(rubyClassName + scriptBase + ":_mouse_exited}");
    };
    return;
  }

  public void setup() {
    if (availableSetup) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_setup");
    }
  }

  public void draw() {
    if (availableDraw) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_draw");
    }
  }

  public void keyPressed() {
    if (availableKeyPressed) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_key_pressed");
    }
  }

  public void keyReleased() {
    if (availableKeyReleased) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_key_released");
    }
  }

  public void mousePressed() {
    if (availableMousePressed) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_mouse_pressed");
    }
  }

  public void mouseReleased() {
    if (availableMouseReleased) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_mouse_released");
    }
  }

  public void mouseClicked() {
    if (availableMouseClicked) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_mouse_clicked");
    }
  }

  public void mouseDragged() {
    if (availableMouseDragged) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_mouse_dragged");
    }
  }

  public void mouseMoved() {
    if (availableMouseMoved) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_mouse_moved");
    }
  }

  public void mouseEntered() {
    if (availableMouseEntered) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_mouse_entered");
    }
  }

  public void mouseExited() {
    if (availableMouseExited) {
      org.ruboto.JRubyAdapter.runRubyMethod(scriptInfo.getRubyInstance(), "send", "_mouse_exited");
    }
  }
}

File: src/processing_sketch_activity.rb
java_import 'com.hoshi.rubotoprocessing2.ProcessingSketchActivity'
require 'forwardable'
require_relative './sketch'

class ProcessingSketchActivity
  import org.ruboto.Log
  include Sketch

  # unused
  # PROTECTED_FIELDS = [
  #   :surfaceReady,
  #   :surfaceChanged,
  #   :dmouseX,
  #   :dmouseY,
  #   :emouseX,
  #   :emouseY,
  #   :windowFocused,
  #   :viewFocused,
  #   :frameRateLastNanos,
  #   :frameRateTarget,
  #   :frameRatePeriod,
  #   :looping,
  #   :redraw,
  #   :paused,
  #   :exitCalled,
  # ]
  PUBLIC_FIELDS = [
    :g,
    :displayWidth,
    :displayHeight,
    #:pixels,       # delegate
    :width,
    :height,
    :mouseX,
    :mouseY,
    :pmouseX,
    :pmouseY,
    :key,
    :keyCode,
    :focused,
    :frameCount,
    :finished,
  ]
  DELEGATE_METHODS = [
    :size,
    :can_draw,
    :canDraw,
    :no_loop,
    :noLoop,
    :pixels,
    :pixels=,
    :load_pixels,
    :update_pixels,
    :smooth,
    :load_font,
    :load_image,
    :load_table,
    :load_xml,
    :load_strings,
    :load_shape,
    :load_bytes,
  ]

  ### helper methods >>> ###
  field_accessor :mousePressed => :field_mouse_pressed
  field_accessor :keyPressed   => :field_key_pressed

  def mouse_pressed?
    @ruboto_java_instance.field_mouse_pressed
  end

  def key_pressed?
    @ruboto_java_instance.field_key_pressed
  end

  def ruboto_log
    Log
  end
  ### <<< helper methods ###

  private

  def synchronize_variables
    PUBLIC_FIELDS.each do |field_name|
      self.send("#{field_name}=", @ruboto_java_instance.send(field_name))
    end
  end

  def check_sketch_method
    [
      :setup,
      :draw,
      :key_pressed,
      :key_released,
      :mouse_pressed,
      :mouse_released,
      :mouse_clicked,
      :mouse_dragged,
      :mouse_moved,
      :mouse_entered,
      :mouse_exited,
    ].each do |m|
      instance_variable_set("@available_#{m}", Sketch.instance_methods(false).include?(m))
    end
  end

  def _setup
    extend SingleForwardable
    def_delegators(:@ruboto_java_instance, *DELEGATE_METHODS)
    check_sketch_method
    synchronize_variables
    setup if @available_setup
  rescue => e
    Log.e "exception in ruby code `setup`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end

  def _draw
    synchronize_variables
    draw if @available_draw
  rescue => e
    Log.e "exception in ruby code `draw`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end

  def _key_pressed
    key_pressed if @available_key_pressed
  rescue => e
    Log.e "exception in ruby code `key_pressed`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end

  def _key_released
    key_released if @available_key_released
  rescue => e
    Log.e "exception in ruby code `key_released`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end

  def _mouse_pressed
    mouse_pressed if @available_mouse_pressed
  rescue => e
    Log.e "exception in ruby code `mouse_pressed`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end

  def _mouse_released
    mouse_released if @available_mouse_released
  rescue => e
    Log.e "exception in ruby code `mouse_released`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end

  def _mouse_clicked
    mouse_clicked if @available_mouse_clicked
  rescue => e
    Log.e "exception in ruby code `mouse_clicked`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end

  def _mouse_dragged
    mouse_dragged if @available_mouse_dragged
  rescue => e
    Log.e "exception in ruby code `mouse_dragged`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end

  def _mouse_moved
    mouse_moved if @available_mouse_moved
  rescue => e
    Log.e "exception in ruby code `mouse_moved`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end

  def _mouse_entered
    mouse_entered if @available_mouse_entered
  rescue => e
    Log.e "exception in ruby code `mouse_entered`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end

  def _mouse_exited
    mouse_exited if @available_mouse_exited
  rescue => e
    Log.e "exception in ruby code `mouse_exited`: #{e}"
    Log.e e.backtrace.join("\n")
    raise e
  end
end
